lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1172 @@
+		mlir::rlc::ConstraintsAnalysis* analysis, 
+		const mlir::Value& result){
+
+		// Keep operand aliases for clarity
+		const auto& op1=op->getOperand(0);
+		const auto& op2=op->getOperand(1);
+		const auto& op1_defop=op1.getDefiningOp();
+		const auto& op2_defop=op2.getDefiningOp();
+
+		// Understand the operand that I have to update
+		mlir::Value update=nullptr;
+		mlir::Value  other=nullptr;
+
+		// Check which operand should be updated (the other should be a constant)
+		// TODO: understand what to do when there are no constants
+		if(op1_defop == nullptr or not mlir::isa<mlir::rlc::Constant>(op1_defop)){
+			update=op1;
+			other=op2;
+		}
+		else if(op2_defop == nullptr or not mlir::isa<mlir::rlc::Constant>(op2_defop)){
+			update=op2;
+			other=op1;
+		}
+		else return false;
+
+		bool ret1=mlir::rlc::updateArithmeticRange(op, currentLattice, analysis, result, update, other, true);
+		bool ret2=mlir::rlc::updateArithmeticRange(op, currentLattice, analysis, result, update, other, false);
+
+		return ret1 or ret2;
+	}
+
+	// Method that checks the lattice and updates the value accordingly
+	static bool updateArithmeticRange(
+		mlir::Operation* op,
+		mlir::rlc::ConstraintsLattice* currentLattice,
+		mlir::rlc::ConstraintsAnalysis* analysis, 
+		const mlir::Value& result,
+		const mlir::Value& update,
+		const mlir::Value& other,
+		bool branch)
+	{
+
+		if (not currentLattice->isValueAValidRange(result, branch)) return false;
+
+		auto result_range = currentLattice->getRange(result,branch);
+
+		ConstraintsLattice::IntegerRange new_range(ConstraintsLattice::APInteger(1),ConstraintsLattice::APInteger(0));
+		// If the other is a constant take its constant value
+		if(auto other_casted=mlir::dyn_cast<mlir::rlc::Constant>(other.getDefiningOp())){
+			auto other_const=other_casted->getAttr("value").dyn_cast<IntegerAttr>().getInt();
+			// Kinda ugly but this is what we have to do with constant ranges
+			new_range=currentLattice->createRange(other_const);
+		}
+		// Else check if the value carried by the operation is 
+		else {
+			// If I have altready a range then i can modify it
+			if(not currentLattice->isValueNotEndRange(other,branch)) return false;
+			// NB: this can become TOP
+			new_range=currentLattice->getRange(other,branch);
+		}
+		// Update only if I found the new range
+		if(new_range.getLower().sle(new_range.getUpper())){
+			// TODO: add more cases of course
+			// TODO: with the llvm class this by default wraps around, do not do that thx
+			llvm::TypeSwitch<mlir::Operation*>(op)
+			.Case<mlir::rlc::AddOp>([&](mlir::rlc::AddOp addop){
+				new_range=result_range.subWithNoWrap(new_range,0,llvm::ConstantRange::PreferredRangeType::Signed);
+			});
+			// Insert the range where is needed
+			return currentLattice->insertOrUpdate(std::make_pair(update,branch),new_range);
+		}
+		// Else case just do nothing
+		return false;
+	}
+
+	// Method called for visiting the yield operation
+	static bool visitYield(
```

Comment:
why is visitYield called both by the interface implementation of yield, and by other places too? Can we rewrite all the users so that they only use the interface method, and then we squash this free function into the method itself?